test/run_test.py

```diff
@@ -340,6 +349,101 @@ def print_to_stderr(message):
+        day_before_commit = str(commit_date - timedelta(days=1)).split(' ')[0]
+        # something like git rev-list --before="2021-03-04" --max-count=1 --remotes="*origin/nightly"
+        nightly_commit = subprocess.check_output(
+            ["git", "rev-list", f"--before={day_before_commit}", "--max-count=1", "--remotes=*origin/nightly"],
+            encoding="ascii").strip()
+        print(f'Using nightly commit: {nightly_commit}')
+
+        job = os.environ.get("CIRCLE_JOB", "")
+        job_minus_shard_number = job.rstrip('0123456789')
+        if not HAVE_BOTO3:
+            print('Please install boto3 to enable automatic test sharding.')
+            return []
+        s3 = boto3.resource("s3", config=botocore.config.Config(signature_version=botocore.UNSIGNED))
+        bucket = s3.Bucket(name="ossci-metrics")
+        summaries = bucket.objects.filter(Prefix=f"test_time/{nightly_commit}/{job_minus_shard_number}")
+    except (RuntimeError, RuntimeWarning):
+        print('Failed to read from S3. Proceeding with no reports.')
+        return []
+
+    reports = []
+    for summary in summaries:
+        binary = summary.get()["Body"].read()
+        string = bz2.decompress(binary).decode("utf-8")
+        reports.append(json.loads(string))
+    return reports
+
+
+def calculate_job_times(reports: List[Dict[str, Any]]) -> Dict[str, Tuple[float, int]]:
+    # an entry will be like ("test_file_name" -> (current_avg, # values))
+    jobs_to_times: Dict[str, Tuple[float, int]] = dict()
+    for report in reports:
+        assert report.get('format_version') == 2, "S3 format currently handled is version 2 only"
+        files: Dict[str, Any] = report['files']
+        for name, test_file in files.items():
+            if name not in jobs_to_times:
+                jobs_to_times[name] = (test_file['total_seconds'], 1)
+            else:
+                curr_avg, curr_count = jobs_to_times[name]
+                new_count = curr_count + 1
+                new_avg = (curr_avg * curr_count + test_file['total_seconds']) / new_count
+                jobs_to_times[name] = (new_avg, new_count)
+    return jobs_to_times
+
+
+def calculate_shards(num_shards: int, tests: List[str], job_times: Dict[str, Tuple[float, int]]) -> List[Tuple[float, List[str]]]:
+    # if there's 'test_cpp_extensions_aot' entry in job_times, add 'test_cpp_extensions_aot_ninja' 
+    # and 'test_cpp_extensions_aot_no_ninja' duplicate entries to ease future computation since 
+    # test_cpp_extensions_aot_no_ninja and test_cpp_extensions_aot_ninja are Python test jobs that 
+    # both use the test_cpp_extensions_aot.py file.
+    if 'test_cpp_extensions_aot' in job_times:
+        job_times['test_cpp_extensions_aot_ninja'] = job_times['test_cpp_extensions_aot']
+        job_times['test_cpp_extensions_aot_no_ninja'] = job_times['test_cpp_extensions_aot']
+    known_job_times: Dict[str, float] = dict()
+    unknown_jobs: List[str] = []
```

Comment:
IMO we can do more simpler by appending 0 to unknown job avg_time.
Since you are doing a reverse sorting and greedy partitioning. all those unknown jobs are just append to the tail of each shard. 
```
adjusted_job_times = {test:job_times.get(test, 0) for test in tests}
```
then you won't need the special treatment for unknown jobs